alpa/collective/collective_group/xla_nccl_collective_group.py

```diff
@@ -0,0 +1,393 @@
+        root_rank = 0
+
+        key = broadcast_options.comm_key
+        comms = self._get_nccl_broadcast_communicator(
+            key, broadcast_options.world_size, broadcast_options.devices_ids,
+            broadcast_options.devices_global_rank)
+        xe.nccl_broadcast_partial_gpus(comms, tensors,
+                                       broadcast_options.local_start_pos_list,
+                                       broadcast_options.n_elements, root_rank)
+
+    def _get_nccl_broadcast_communicator(self,
+                                         comm_key,
+                                         world_size,
+                                         devices_ids,
+                                         devices_global_rank,
+                                         nccl_uid=None):
+        """Create or retrieve a list of NCCL communicators for
+        broadcast from cache. Here we only use partial devices in a host, so
+        we create this function besides _get_nccl_collective_communicator.
+
+        If the communicator is found in cache, return the communicator. If not,
+        a communicator and a stream will be created and put in cache.
+
+        Args:
+            comm_key (str): the key to query the communicator cache.
+            world_size (int): the number of devices in this collective
+                              communicator.
+            devices_ids (List): a list of GPU devices of the current process
+                                that participates into the collective.
+            devices_global_rank (List): the corresponding global rank for
+                                device in devices_ids.
+            nccl_uid : If it is None, we will create a nccl_uid here.
+
+        Returns:
+            communicator: the NCCL communicator corresponded to the devices.
+        """
+        if not comm_key:
+            raise RuntimeError("Got empty communicator key.")
+
+        for d in devices_ids:
+            self._used_gpu_indices.add(d)
+
+        # TODO(Hao): lock the _dev_comm_map here.
+        if comm_key in self._dev_comm_map:
+            return self._dev_comm_map[comm_key]
+
+        group_key = self._generate_group_key(comm_key)
+        if devices_global_rank[0] == 0:
+            if nccl_uid is None:
+                nccl_uid = self._generate_nccl_uid(group_key)
+        else:
+            if nccl_uid is None:
+                rendezvous = Rendezvous(group_key)
+                rendezvous.meet()
+                nccl_uid = rendezvous.get_nccl_id()
+
+                # Recycle the NCCLUniqueIDStore named actor *pro-activately* to
+                # avoid named actor leak.
+                if rendezvous.get_access_counter() == self.world_size:
+                    logger.debug(
+                        "NCCLUniqueID has been broadcasted. The "
+                        "NCCLUniqueIDStore will go out of context and be "
+                        "destroyed.")
+                    rendezvous.destroy_store()
```

Comment:
Again, avoid copying codes and consider moving the shared part into a common base class.